docs/SILProgrammersManual.md

```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+
+This example has an identifiable def-use chain from `%base` to `%access`:
+```
+bb0:
+  %base = alloc_box $Int { var Int }
+  %boxadr = project_box %base : ${ var Int }
+  %p0 = address_to_pointer %boxadr : $*Int to $Builtin.RawPointer
+  cond_br _, bb1, bb2
+
+bb1:
+  %p1 = copy_value %p0 : $Builtin.RawPointer
+  br bb3(%p1 : $Builtin.RawPointer)
+
+bb2:
+  br bb3(%p0 : $Builtin.RawPointer)
+
+bb3(%ptr : $Builtin.RawPointer):
+  %adr = pointer_to_address %ptr : $Builtin.RawPointer to $*Int
+  %access = begin_access [read] [static] %adr : $*Int
+```
+
+Note that address-type phis are illegal (full enforcement
+pending). This is important for simplicity and efficiency, but also
+allows for a class of storage optimizations, such as bitfields, in
+which address storage is always uniquely determined. Currently, if a
+(non-address) phi on the access path from `base` to `access` does not
+have a common base, then it is considered an invalid access (the
+AccessedStorage object is not valid). SIL verification ensures that a
+formal access always has valid AccessedStorage (WIP). In other words, the
+source of a `begin_access` marker must be a single, non-phi base. In
+the future, for further simplicity, we may generally disallow box and
+pointer phis unless they have a common base.
+
+Not all SIL memory access is part of a formal access, but the
+`AccessedStorage` and `AccessPath` abstractions are universally
+applicable. Non-formal access still has an access base, even though
+the use-def search does not begin at a `begin_access` marker. For
+non-formal access, SIL verification is not as strict. An invalid
+access is allowed, but handled conservatively. This is safe as long as
+those non-formal accesses can never alias with class and global
+storage. Class and global access is always guarded by formal access
+markers--at least until static markers are stripped from SIL.
+
+#### Nested access
+
+Nested access occurs when an access base is a function argument. The
+caller always checks `@inout` arguments for exclusivity. However, the
+argument itself is a variable with its own formal access. Conflicts
+may occur in the callee which were not evident in the caller:
+
+```
+sil @takesTwoInouts : $@convention(thin) (@inout Int, @inout Int) -> ()
+
+sil @hasNestedAccess : $@convention(thin) (@inout Int) -> () {
+bb0(%0 : $*Int):
+  %innerAccess  = begin_access [modify] %0 : $*Int
+  %conflicting  = begin_access [modify] %0 : $*Int
+  %f = function_ref @takesTwoInouts
+  apply %f(%innerAccess, %conflicting)
+    : $@convention(thin) (@inout Int, @inout Int) -> ()
+  end_access %conflicting : $*Int
+  end_access %innerAccess : $*Int
+  //...
+}
+
+%var = alloc_stack $Int
+%outerAccess  = begin_access [modify] %var : $*Int
+%f = function_ref @hasNestedAccess
+apply %f(%outerAccess) : $@convention(thin) (@inout Int) -> () {
+end_access %outerAccess : $*Int
+```
+
+Nested accesses become part if the def-use chain after inlining.
```

Comment:
Relevant updated sections:
```
Nested accesses become part if the def-use chain after inlining. Here,
both `%innerAccess` and `%conflicting` are nested within
`%outerAccess`:
```

```
%var = alloc_stack $Int
%outerAccess  = begin_access [modify] %var : $*Int
%innerAccess  = begin_access [modify] %outerAccess : $*Int
%conflicting  = begin_access [modify] %outerAccess : $*Int
%f = function_ref @takesTwoInouts
apply %f(%innerAccess, %conflicting)
  : $@convention(thin) (@inout Int, @inout Int) -> ()
end_access %conflicting : $*Int
end_access %innerAccess : $*Int
end_access %outerAccess : $*Int
```

```
For most purposes, the inner access scopes are irrelevant. When we ask
for the "accessed storage" for `%innerAccess`, we get an
`AccessedStorage` value of "Stack" kind with base `%var =
alloc_stack`. If instead of finding the original accessed storage, we
want to identify the enclosing formal access scope, we need to use a
different API that supports the special `Nested` storage kind. This is
typically only used for exclusivity diagnostics though.

TODO: Nested static accesses that result from inlining could
potentially be removed, as long as DiagnoseStaticExclusivity has
already run.

#### Access projections

On the def-use chain between the *outermost* formal access scope within
the current function and a memory operation, *access projections*
identify subobjects laid out within the formally accessed
variable. The sequence of access projections between the base and the
memory address correspond to an access path.

For example, there is no formal access for struct fields. Instead,
they are addressed using a `struct_element_addr` within the access
scope:
```

```
%access  = begin_access [read] [static] %base : $*S
%memaddr = struct_element_addr %access : $*S, #.field
%value   = load [trivial] %memaddr : $*Int64
end_access %access : $*S
```

```
Note that is is possible to have a nested access scope on the address
of a struct field, which may show up as an access of
struct_element_addr after inlining.
```